hazelcast-tpc-engine/src/main/java/com/hazelcast/internal/tpc/buffer/ThreadLocalBuffer.java

```diff
@@ -0,0 +1,288 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.tpc.buffer;
+
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+
+import static com.hazelcast.internal.tpc.nio.Bits.INT_SIZE_IN_BYTES;
+import static com.hazelcast.internal.tpc.nio.Bits.LONG_SIZE_IN_BYTES;
+import static com.hazelcast.internal.tpc.nio.Bits.SHORT_SIZE_IN_BYTES;
+
+public class ThreadLocalBuffer implements Buffer {
+    ByteBuffer[] chunks;
+
+    private final ConcurrentBufferAllocator superAllocator;
+    private ThreadLocalBufferAllocator allocator;
+
+    /**
+     * Position of a last byte in a buffer.
+     */
+    private int pos;
+
+    private int limit;
+
+    /*&
+     * Position of a last chunk in.
+     */
+    private int chunksPos;
+
+    private int chunkToRelease;
+
+    public ThreadLocalBuffer(ThreadLocalBufferAllocator allocator, int minSize, ConcurrentBufferAllocator concurrentAllocator) {
+        this.allocator = allocator;
+        this.superAllocator = concurrentAllocator;
+        this.chunks = new ByteBuffer[((minSize - 1) / ThreadLocalBufferAllocator.BUFFER_SIZE) + 1];
+
+        for (int i = 0; i < chunks.length; i++) {
+            addChunk(allocator.getNextByteBuffer());
+        }
+    }
+
+    void reset(int minSize) {
+        this.chunksPos = 0;
+        this.pos = 0;
+        this.limit = 0;
+        this.chunkToRelease = 0;
+
+        for (int i = 0; i < ((minSize - 1) / ThreadLocalBufferAllocator.BUFFER_SIZE) + 1; i++) {
+            addChunk(allocator.getNextByteBuffer());
+        }
+    }
+
+    void overtakenBy(ThreadLocalBufferAllocator allocator) {
+        this.allocator = allocator;
+    }
+
+    @Override
+    public void release() {
+        if (superAllocator == null) {
+            allocator.free(this);
+            return;
+        }
+        superAllocator.free(this);
+    }
+
+    @Override
+    public void releaseNextChunk(ByteBuffer chunk) {
+        assert chunk == chunks[chunkToRelease];
+
+        if (superAllocator == null) {
+            allocator.free(chunks[chunkToRelease++]);
+            return;
+        }
+        superAllocator.free(chunks[chunkToRelease++]);
+    }
+
+    @Override
+    public int position() {
+        return pos;
+    }
+
+    @Override
+    public void clear() {
+        for (int i = 0; i < chunksPos; i++) {
+            chunks[i].clear();
+        }
+        pos = 0;
+    }
+
+    @Override
+    public void flip() {
+        limit = pos;
+        pos = 0;
+        for (int i = 0; i < chunksPos; i++) {
+            chunks[i].flip();
+        }
+    }
+
+    @Override
+    public byte getByte(int pos) {
+        int chunk = pos / ThreadLocalBufferAllocator.BUFFER_SIZE;
+        int posInChunk = pos % ThreadLocalBufferAllocator.BUFFER_SIZE;
+        return chunks[chunk].get(posInChunk);
+    }
+
+    @Override
+    public void writeByte(byte src) {
```

Comment:
Maybe we can collect all these write methods at the end and separate them with the rest of the class. Something like [this](https://github.com/hazelcast/hazelcast/blob/2813d007a6b7d168abf8f6fe9557712c69601dfb/hazelcast/src/test/java/com/hazelcast/test/HazelcastTestSupport.java#L242-L244).